app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt

```diff
@@ -84,32 +112,201 @@ class StateFragmentPresenter @Inject constructor(
+    logger.d("StateFragment", "hasPreviousState: $hasPreviousState")
+    logger.d("StateFragment", "hasNextState: $hasNextState")
+    stateViewModel.hideAllButtons()
+    stateViewModel.setPreviousButtonVisible(hasPreviousState)
+    if (!hasNextState) {
+      when (interactionId) {
+        CONTINUE -> stateViewModel.setContinueButtonVisible(true)
+        END_EXPLORATION -> stateViewModel.setEndExplorationButtonVisible(true)
+        FRACTION_INPUT -> stateViewModel.setSubmitButtonVisible(true)
+        ITEM_SELECT_INPUT -> stateViewModel.setSubmitButtonVisible(true)
+        MULTIPLE_CHOICE_INPUT -> stateViewModel.setSubmitButtonVisible(true)
+        NUMERIC_INPUT -> stateViewModel.setSubmitButtonVisible(true)
+        NUMERIC_WITH_UNITS -> stateViewModel.setSubmitButtonVisible(true)
+        TEXT_INPUT -> stateViewModel.setSubmitButtonVisible(true)
+      }
+    } else {
+      stateViewModel.setNextButtonVisible(hasNextState)
+    }
   }
 
+  /**
+   * This function subscribes to current state of the exploration.
+   * Whenever the current state changes it will automatically get called and therefore their is no need to call this separately.
+   * This function currently provides important information to decide which button we should display.
+   */
   private fun subscribeToCurrentState() {
-    ephemeralStateLiveData.observe(fragment, Observer<EphemeralState> { result ->
-      logger.d("StateFragment", "getCurrentState: ${result.state.name}")
+    ephemeralStateLiveData.observe(fragment, Observer<EphemeralState> { it ->
+      resultEphemeralState.set(it)
+      logger.d("StateFragment", "stateName: ${it.state.name}")
+      val interactionId = it.state.interaction.id
+      val hasPreviousState = it.hasPreviousState
+      val hasNextState = it.completedState.answerCount > 0
+      controlButtonVisibility(interactionId, hasPreviousState, hasNextState)
     })
   }
 
+  /** Helper for subscribeToCurrentState. */
   private val ephemeralStateLiveData: LiveData<EphemeralState> by lazy {
     getEphemeralState()
   }
 
+  /** Helper for subscribeToCurrentState. */
   private fun getEphemeralState(): LiveData<EphemeralState> {
     return Transformations.map(explorationProgressController.getCurrentState(), ::processCurrentState)
   }
 
+  /** Helper for subscribeToCurrentState. */
   private fun processCurrentState(ephemeralStateResult: AsyncResult<EphemeralState>): EphemeralState {
     if (ephemeralStateResult.isFailure()) {
       logger.e("StateFragment", "Failed to retrieve ephemeral state", ephemeralStateResult.getErrorOrNull()!!)
     }
     return ephemeralStateResult.getOrDefault(EphemeralState.getDefaultInstance())
   }
+
+  /**
+   * This function listens to the result of submitAnswer.
+   * Whenever an answer is submitted using ExplorationProgressController.submitAnswer function,
+   *  this function will wait for the response from that function and based on which we can move to next state.
```

Comment:
Nit: one space between ``*`` and ``this``

Reply:
Ditto elsewhere.

Reply:
Okay resolved that.